### PR TITLE
Feat/factory

### PR DESCRIPTION
**Pull request by T04435**
Add a `RoundFactory` contract to Jubi's ecosystem to manage round creation.

This will be followed up by a PR adding ACL to provide better management of ventures/rounds more details #23

### CODE REVIEW

File: contracts/Presale.sol [hunk 1]
```diff
@@ -64,8 +64,12 @@ contract Presale is Ownable {
         uint256 _vestingCliffDuration,
         uint256 _vestingDuration,
         IERC20 _raiseToken,
-        address _daoMultisig
+        address _daoMultisig,
+        bool factory
     ) {
+        if (factory) {
+            _transferOwnership(_daoMultisig);
+        }
```
Comment (T04435) [inline contracts/Presale.sol:72]: I could be missing something here.

IMHO we need to transfer ownership to DAO_MSig, if not the `JubiFactory` would be the owner and the only way to interact newly deployed contract would be to call it with `JubiFactory`

File: test/jubi-factory-tests.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        it("should only allow owner to close round", async () => {
+          await expect(PRESALE.connect(VENTURE2).closeRound())
+            .to.revertedWith(ONLY_OWNER_ERROR);
+
+          await PRESALE.connect(VENTURE1).closeRound();
+          expect(await PRESALE.isOpen()).false;
+        });
+      });
+
+      describe("Round Interactions", () => {
```
Comment (T04435) [inline test/jubi-factory-tests.ts:236]: Should I replicate all test from `presale-test.ts` here?

File: contracts/Presale.sol (hunk 1 above)
Comment (jsuresh) [reply contracts/Presale.sol:72]: THis transfer should live in the factory, not to _daoMultisig but to the caller of createRound

File: contracts/JubiFactory.sol [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    function createRound(PresaleConfig memory config) external onlyOwner {
+        uint256 hardCap = config.hardCap;
+        uint256 hurdle = config.hurdle;
+        bytes32 inviteCodesMerkleRoot = config.inviteCodesMerkleRoot;
+        uint256 vestingCliffDuration = config.vestingCliffDuration;
+        uint256 vestingDuration = config.vestingDuration;
+        IERC20 raiseToken = config.raiseToken;
+        address daoMultisig = config.daoMultisig;
+
+        Presale Round = new Presale(hardCap, hurdle, inviteCodesMerkleRoot, vestingCliffDuration, vestingDuration, raiseToken, daoMultisig, true);
```
Comment (jsuresh) [inline contracts/JubiFactory.sol:42]:
```suggestion
        Presale Round = new Presale(hardCap, hurdle, inviteCodesMerkleRoot, vestingCliffDuration, vestingDuration, raiseToken, daoMultisig, true);
        round.transferOwnership(msg.sender)
```

File: contracts/JubiFactory.sol [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+        ventureRounds[daoMultisig].push(Round);
```
Comment (jsuresh) [inline contracts/JubiFactory.sol:44]: THis implies anyone can create a round for a given daoMultisig

daoMultisig IMHO is a bad name :) I was tired and it's taking a life of it's own.

It's the account funding ends up in on round close, if it passes the hurdle (or any other pre-conditions we decide to set).

For now, make this `msg.sender` - we'll need a concept of a venture for which I've created an issue where we can discuss/shape (not required immediately)

Comment (jsuresh) [reply contracts/JubiFactory.sol:44]: https://github.com/JubiDAO/protocol/issues/23

File: contracts/JubiFactory.sol [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+        emit RoundCreated(daoMultisig, Round);
+    }
+
+    /// @notice Retrieves the list of rounds for the Venture: `venture`
+    function getRounds(address venture) public view returns (bytes[] memory rounds) {
```
Comment (jsuresh) [inline contracts/JubiFactory.sol:50]: Cool, way to get it all in a single query :)

Worth noting this does add extra data on-chain, which, is always worth optimising for _however_ this contract is only deployed once, and this looks like it makes the FE much simpler!

ps - `bytes32` over bytes as we know it's size given it's an on chain address.

File: contracts/JubiFactory.sol [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./Presale.sol";
+
+/**
+ * @title Contract to Create Jubi's Presale Contract
+ * @author @T04435
+ * @notice You can use this Contract to create new Instances of Jubi Presale Contracts `Presale.sol`
+ */
+contract JubiFactory is Ownable {
```
Comment (jsuresh) [inline contracts/JubiFactory.sol:12]: s/JubiFactory/RoundFactory/g

Comment (T04435) [reply contracts/JubiFactory.sol:12]: idk vim :)))

File: contracts/Presale.sol (hunk 1 above)
Comment (T04435) [reply contracts/Presale.sol]: updated.

File: contracts/JubiFactory.sol (hunk 3 above)
Comment (T04435) [reply contracts/JubiFactory.sol]: updated

File: test/round-factory-tests.ts [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            toAtto(100),
+            ...nextInvite()
+          );
+          await presale.depositFor(
+            ash_address,
+            toAtto(500),
+            ...nextInvite()
+          );
+
+          expect(await presale.allocation(jeeva_address)).eql(toAtto(100));
```
Comment (ashvds) [inline test/round-factory-tests.ts:235]: Going with real life examples?

File: contracts/JubiFactory.sol (hunk 5 above)
Comment (T04435) [reply contracts/JubiFactory.sol]: as discussed no need to update `bytes => bytes32`.

Comment (ashvds) [reply contracts/JubiFactory.sol]: @T04435 , if the intent is that we deploy this contract only once, then should we have the ability to update the composeable bits that it uses? or is this contract updateable?

File: test/round-factory-tests.ts (hunk 7 above)
Comment (samfroggy) [reply test/round-factory-tests.ts:235]: This is always the way, right? `account1` is boring. Jeeva and Ash are way cooler

File: contracts/JubiFactory.sol (hunk 5 above)
Comment (samfroggy) [reply contracts/JubiFactory.sol]: I will write a subgraph to index all the information for any given params, i.e the `msg.sender` of the `createRound` or the `_daoMultisig` for a round.

We then can expand to withdrawal events on the presale, so on and so forth for pretty front-end things

**Review by samfroggy — DISMISSED** [review]

**Review by samfroggy — APPROVED** [review]

**Pull request merged**